### ISSUE TITLE
Try Python script to create job submission script from RiboViz configuration

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#289]
Please try out the Python script to customise a template job submission script and note your comments here.

See updated documentation in:

https://github.com/riboviz/riboviz/blob/develop/docs/user/run-on-eddie.md

and

https://github.com/riboviz/riboviz/blob/develop/docs/user/create-job-script.md

**Comment by ewallace** [comment]
Tried this with `riboviz/riboviz` develop branch at commit d040346, simultaneously testing [example-datasets#65](https://github.com/riboviz/example-datasets/pull/65).

First I tried the validate only with smaller memory footprint:

```
python -m riboviz.tools.create_job_script \
    -i jobs/eddie-template.sh \
    -o job_riboviz_Lareau_2014_validate.sh \
    --config-file ../example-datasets/fungi/saccharomyces/Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml \
    --r-libs /exports/csce/eddie/biology/groups/wallace_rna/Rlibrary \
    --job-runtime "00:05:00" \
    --job_memory 1G \
    --job_num_cpus 1 \
    --validate-only
```

This led to an error in the arguments:
```
create_job_script.py: error: unrecognized arguments: --job_memory 1G --job_num_cpus 1
```

That indicates a bug; as of this commit, `--job_memory` and `--job_num_cpus`flags aren't working; they're essential.

So I next tried to validate with default memory/cpus:

```
python -m riboviz.tools.create_job_script \
    -i jobs/eddie-template.sh \
    -o job_riboviz_Lareau_2014_validate.sh \
    --config-file ../example-datasets/fungi/saccharomyces/Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml \
    --r-libs /exports/csce/eddie/biology/groups/wallace_rna/Rlibrary \
    --job-runtime "00:05:00" \
    --validate-only
```

Then submitted this using:
```
qsub -q *@@ringfence_wallace job_riboviz_Lareau_2014_validate.sh
```

That appears to be syntactically correct, yet did not actually run; it was stuck on the cluster awaiting submission for over a week. Asking for information on the job with `qstat -j 10329347` seemed to indicate a scheduling problem:

```
scheduling info:            queue instance "gpu@node1j17.ecdf.ed.ac.uk" dropped because it is temporarily not available (similar reason #1)
[...]
                            queue instance "eddie@node1H17.ecdf.ed.ac.uk" dropped because it is full (similar reason #229)
                            cannot run in PE "mpi" because resource requirements of the job cannot be fulfilled

```

I do not know if this is a problem with the `PE "mpi"` itself, or with the resource requirements, or with the queue.

Next I created the job submission script:
```
python -m riboviz.tools.create_job_script \
    -i jobs/eddie-template.sh \
    -o job_riboviz_Lareau_2014.sh \
    --config-file ../example-datasets/fungi/saccharomyces/Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml \
    --r-libs /exports/csce/eddie/biology/groups/wallace_rna/Rlibrary \
    --job-runtime "02:00:00"
```

At this point I gave up. Any ideas?

**Comment by mikej888** [comment]
There are typos in your first invocation. The command line arguments `--job-memory` and `--job-num-cpus` should  be dash delimited not underscore delimited, see
https://github.com/riboviz/riboviz/blob/develop/docs/user/create-job-script.md.

The others seem to be related to the job resource requested rather than the script. I couldnt get an `mpi` job to run, it just sat in the queue until I deleted it, so I used `--job-parallel-env sharedmem`.

**Comment by ewallace** [comment]
Sorry about my typos!

Let's change the default to `--job-parallel-env sharedmem`.

**Comment by ewallace** [comment]
I'll try to fix those in `edit-job-sub-289` branch

**Comment by ewallace** [comment]
Would it make sense to change the `input_file` and `output` parameters to `input-file` and `output-file`, for consistency?

**Comment by mikej888** [comment]
Yes, apologies that inconsistency slipped in.

**Comment by mikej888** [comment]
@ewallace, sorry I, and the doc, should have beem more clear.

The template job script is configured by parameters in a RiboViz YAML configuration file and/or any command-line arguments provided to `riboviz.tools.create-job-script`

The table of parameters in `create-job-script.md` refer to the YAML configuration parameters (it is a subset of those in `docs/user/prep-riboviz-config.md`). These need to be underscore-delimited for consistency with all our other configuration parameters and for consistency with the tokens in the template job submission scripts (`jobs/eddie-template.sh`). Dash-delimited configuration parameters in the YAML will be ignored. The changes you made to the table need to be reverted.

The complementary command-line parameters are dash-delimited for consistency with how such verbose parameters are often specified at the bash command-line.

If you'd prefer consistency between the command-line tool parameters and the YAML file, then make the following updates to use underscore-delimited parameters on the command-line e.g. (`... --job_runtime "02:00:00" ...`):

* Edit `riboviz.tools.create_job_config` and, in the code that sets up the command-line parameters, replace `param.replace("_", "-")` with `param`; replace `--token-tag` with `--token_tag`, replace `--input_file` with `input` (this was an oversight anyway as the output file has long parameter `output`); update the Usage in the comment at the top of the file.  
* Edit `docs/user/create-job-script.md`, update the Usage.
* Edit `docs/user/run-on-eddie.md`, update the `riboviz.tools.create-job.script` example.

**Comment by mikej888** [comment]
Yes, that's fine.

An additional thought is that, adopting underscore delimited command line parameters incurs far less overhead than moving to dash delimited paraneters in all our YAML files (as we should avoid mixing dashes and underscores in our YAML)

**Comment by mikej888** [comment]
Discussion at 24/03/21 meeting decided on option 1) from #296 Use snake_case in create_job_script.

**Comment by 3mma-mack** [comment]
I have tried using create_job_script to create job submission scripts for the Duncan et al _S. pombe_ dataset and the Guydosh _S. cerevisiae_  dataset. Overall it worked well and ran perfectly, with the Duncan dataset producing identical results to those produced previously.

The only problem I ran into was when I mistyped the path for the `--r-libs` parameter, which lead to the error:
```
Error in library(getopt, quietly = T) :
  there is no package called ‘getopt’
Calls: suppressMessages -> withCallingHandlers -> library
Execution halted
```
This occurred at Bam_to_H5, which initially confused me as it was far along in the process, but makes sense if that s the first time an R package is called.
